### PR TITLE
Rebase #14178 / Add not_before_duration API parameter to Root/Intermediate CA generation

### PR DESCRIPTION
**Pull request by cipherboy**
Rebase of #14178 so we can land it in this release cycle :-)

This builds on the excellent work of @guysv with some minor updates:

1. This shouldn't occur on the Intermediate _CSR_ generation, as the validity period isn't conveyed at all on the CSR request and is solely at the discretion of the CA Issuer.
2. Decided to add this to the existing field helper rather than a new one.
3. Needed a test case :-)

Since we changed a lot of the structure of the PKI engine to support multiple issuers, this mostly just needed a rebase to catch the new `buildPath...` method locations. :-)

### CODE REVIEW

File: builtin/logical/pki/fields.go [hunk 1]
```diff
@@ -250,12 +250,22 @@ If you want more than one, specify alternative names in the alt_names
...
 	fields["not_after"] = &framework.FieldSchema{
 		Type: framework.TypeString,
 		Description: `Set the not after field of the certificate with specified date value.
 The value format should be given in UTC format YYYY-MM-ddTHH:MM:SSZ`,
 	}
 
+	fields["not_before_duration"] = &framework.FieldSchema{
+		Type:        framework.TypeDurationSecond,
+		Default:     30,
+		Description: `The duration before now the cert needs to be created / signed.`,
```
Comment (kitography) [inline builtin/logical/pki/fields.go:263]: I don't understand this description.

**Review by kitography — APPROVED** [review]
LGTM

File: builtin/logical/pki/fields.go (hunk 1 above)
Comment (cipherboy) [reply builtin/logical/pki/fields.go:263]: That's actually consistent with the existing description. I think perhaps we should update both... Maybe 

> The duration before now which the certificate needs to be backdated by.

is clearer?

File: builtin/logical/pki/fields.go [hunk 2]
```diff
@@ -250,12 +250,22 @@ If you want more than one, specify alternative names in the alt_names
...
+
 	fields["not_after"] = &framework.FieldSchema{
 		Type: framework.TypeString,
 		Description: `Set the not after field of the certificate with specified date value.
 The value format should be given in UTC format YYYY-MM-ddTHH:MM:SSZ`,
 	}
 
+	fields["not_before_duration"] = &framework.FieldSchema{
+		Type:        framework.TypeDurationSecond,
+		Default:     30,
```
[the remark this answers is not in the document]
Comment (cipherboy) [reply builtin/logical/pki/fields.go:262]: Also, its generally helpful if you issue an intermediate and immediately issue a leaf, the leaf will be backdated by 30s, but your issuer isn't, so if the target system is in that (-30, -1) window that the intermediate isn't valid for, you've gotta wait 30s before the leaf will be considered valid/trusted.

**Review by stevendpclark — APPROVED** [review]

File: builtin/logical/pki/fields.go (hunk 1 above)
Comment (kitography) [reply builtin/logical/pki/fields.go:263]: Yeah, that's much clearer, thank you

**Comment by cipherboy** [comment]
Thanks all!

**Comment by guysv** [comment]
Being lazy pays off 😅 thanks for the merge!!

**Pull request merged**